### PR TITLE
Ensure parametric MCM objects are compatible with Jax-centric angle parameters

### PR DESCRIPTION
**Pull request by mlxd**
**Context:** This PR fixes an issue where Jax-centric angle inputs were unable with the the parametric MCM inputs. The issue was due to hashability of array data, and the lack of unpacking to a fixed tuple.
To avoid failures, the hashing of the underlying data now relies on conversion of the angle to a fixed array, which is then fed into an instance of the [hashlib](https://docs.python.org/3/library/hashlib.html) library's implementations. The underlying numpy array follows the Python buffer API, and so the data is directly readable for a hashed representation. 

**Description of the Change:** As above.

**Benefits:** Ensures the angle parameter is compatible with Python floats, numpy and Jax 0-dim float arrays, and even Jitted functions.

**Possible Drawbacks:** As this isn't anticipated to be used anywhere outside of the MCM functionality, no concerns are currently raised for collisions between hashes of different objects. However, due to the inability to hash JITable objects, the implementation choice for relying on the string angle may 

**Related GitHub Issues:**

### CODE REVIEW

**Comment by mlxd** [comment]
[sc-93471]

File: tests/ftqc/test_parametric_mid_measure.py [hunk 1]
```diff
@@ -539,11 +538,20 @@ def circ():
...
+    ):
         """Test that the measure_ functions are captured as expected"""
         import jax
         import networkx as nx
 
+        if not angle_type == "float":
+            import importlib
+
+            np_type = importlib.import_module(angle_type)
+            angle = np_type.array(angle)
```
Comment (joeycarter) [inline tests/ftqc/test_parametric_mid_measure.py]: Do we need `importlib` here? Why not just

```python
if angle_type == "numpy":
    angle = np.array(angle)
elif angle_type == "jax":
    angle = jax.numpy.array(angle)
```

with the `@pytest.mark.jax` decorator added to the function. (We've already imported `numpy` at the top of the module).

File: tests/ftqc/test_parametric_mid_measure.py [hunk 2]
```diff
@@ -919,22 +931,48 @@ def circ():
...
+            # https://docs.pennylane.ai/en/stable/introduction/dynamic_quantum_circuits.html#tree-traversal-algorithm
+            pytest.skip("TT & jax.jit are incompatible")
+
         dev = qml.device("default.qubit", shots=shots)
 
+        if not angle_type == "float":
+            import importlib
+
+            np_type = importlib.import_module(angle_type)
+            angle = np_type.array(angle)
```
Comment (joeycarter) [inline tests/ftqc/test_parametric_mid_measure.py]: Same question here.

File: tests/ftqc/test_parametric_mid_measure.py [hunk 3]
```diff
@@ -919,22 +931,48 @@ def circ():
             assert np.allclose(circ(), [np.cos(2.345), -1])
 
     @pytest.mark.parametrize("mcm_method, shots", [("tree-traversal", None), ("one-shot", 10000)])
-    def test_diagonalize_mcms_returns_parametrized_mcms(self, mcm_method, shots):
+    @pytest.mark.parametrize("angle", [0.1234, np.array([-0.4321])])
```
Comment (joeycarter) [inline tests/ftqc/test_parametric_mid_measure.py:933]: Maybe a non-issue, but do we need to worry about the cases where the array has shape

* `()`, e.g. `np.array(1.234)`
* `(1, 1)`, e.g. `np.array([[1.234]])`

File: pennylane/ftqc/parametric_midmeasure.py [hunk 4]
```diff
@@ -389,10 +391,20 @@ def _flatten(self):
     @property
     def hash(self):
         """int: Returns an integer hash uniquely representing the measurement process"""
+        if is_abstract(self.angle):  # pragma: no cover
+            # no unique value from tracer to values, hash based on object string
+            param_hash = hashlib.sha256(str(self).encode()).digest()
```
Comment (joeycarter) [inline pennylane/ftqc/parametric_midmeasure.py:396]: Just curious, how do you decide on the appropriate SHA-2 length to use? e.g. `sha256` vs `sha512`?

**Review by joeycarter — APPROVED** [review]
Looks good to me, I only had a couple of questions.

File: pennylane/ftqc/parametric_midmeasure.py (hunk 4 above)
Comment (multiphaseCFD) [inline pennylane/ftqc/parametric_midmeasure.py:396]: Just want to confirm if we want to convert the `self` object or `self.angle` to a `str` object?

File: tests/ftqc/test_parametric_mid_measure.py [hunk 5]
```diff
@@ -499,7 +499,7 @@ def test_measure_z_dispatches_to_measure(self, reset, postselect):
     @pytest.mark.jax
     @pytest.mark.usefixtures("enable_disable_plxpr")
     @pytest.mark.parametrize(
-        "meas_func, angle, plane", [(measure_x, 0.0, "XY"), (measure_y, 1.5707, "XY")]
+        "meas_func, angle, plane", [(measure_x, 0.0, "XY"), (measure_y, np.pi / 2, "XY")]
```
Comment (multiphaseCFD) [inline tests/ftqc/test_parametric_mid_measure.py:502]: Do we want to use `math.pi` or `np.pi` here?

**Review by multiphaseCFD — APPROVED** [review]
Thanks @mlxd ! Looks good to me. Just a couple of questions.

File: tests/ftqc/test_parametric_mid_measure.py (hunk 5 above)
Comment (joeycarter) [reply tests/ftqc/test_parametric_mid_measure.py:502]: They're equivalent, no?

```pycon
>>> import math
>>> import numpy as np
>>> math.pi == np.pi
True
```

File: pennylane/ftqc/parametric_midmeasure.py (hunk 4 above)
Comment (mlxd) [reply pennylane/ftqc/parametric_midmeasure.py:396]: There's no better solution here when dealing with a jax tracer, as there's no numeric value. Using the string offers us the unique memory address as part of the string, which is treated correctly with the buffer protocol of the hasher.

File: tests/ftqc/test_parametric_mid_measure.py (hunk 5 above)
Comment (mlxd) [reply tests/ftqc/test_parametric_mid_measure.py:502]: Numpy is fine for constants

File: pennylane/ftqc/parametric_midmeasure.py (hunk 4 above)
Comment (mlxd) [reply pennylane/ftqc/parametric_midmeasure.py:396]: sha256 was a good balance between performance, and unique-able hashiness (ie low collision likelihood).

File: tests/ftqc/test_parametric_mid_measure.py (hunk 3 above)
Comment (mlxd) [reply tests/ftqc/test_parametric_mid_measure.py:933]: I think it is fine for now. The hashers will handle all of the above

File: tests/ftqc/test_parametric_mid_measure.py (hunk 1 above)
Comment (mlxd) [reply tests/ftqc/test_parametric_mid_measure.py]: I had added it for all of the ML frameworks, but with just NP and JAX, you are right, that's simpler.

File: tests/ftqc/test_parametric_mid_measure.py (hunk 2 above)
Comment (mlxd) [reply tests/ftqc/test_parametric_mid_measure.py]: Yep, same thing here. I'll simplify

**Review by lillian542 — APPROVED** [review]

File: tests/ftqc/test_parametric_mid_measure.py (hunk 2 above)
Comment (mlxd) [reply tests/ftqc/test_parametric_mid_measure.py]: Done

File: tests/ftqc/test_parametric_mid_measure.py (hunk 1 above)
Comment (mlxd) [reply tests/ftqc/test_parametric_mid_measure.py]: Done
